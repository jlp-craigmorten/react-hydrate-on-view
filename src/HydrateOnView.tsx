import { Suspense, useEffect, useRef, useState } from 'react';

import type { FC, PropsWithChildren } from 'react';

interface SuspendConditionallyProps {
  suspend: boolean;
}

const SuspendConditionally: FC<
  PropsWithChildren<SuspendConditionallyProps>
> = ({ suspend, children }) => {
  if (suspend && typeof window !== 'undefined') {
    throw new Promise(() => {});
  }

  return <>{children}</>;
};

export interface HydrateOnViewProps {
  /**
   * Margin around the viewport. Can have values similar to the CSS margin property,
   * e.g. `10px 20px 30px 40px` (top, right, bottom, left).
   */
  rootMargin?: string;
  /**
   * Styles to apply to the container `<div>` element.
   */
  style?: React.CSSProperties;
  /**
   * Number between `0` and `1` indicating the percentage of the content that
   * should be visible before triggering hydration.
   */
  threshold?: number;
}

export const HydrateOnView: FC<PropsWithChildren<HydrateOnViewProps>> = ({
  children,
  rootMargin,
  style,
  threshold,
}) => {
  const ref = useRef<HTMLDivElement>(null);
  const [inView, setInView] = useState(false);

  useEffect(() => {
    if (inView) {
      return;
    }

    const observer = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting) {
          observer.disconnect();
          setInView(true);
        }
      },
      {
        rootMargin,
        threshold,
      },
    );

    observer.observe(ref.current!);

    return () => {
      observer.disconnect();
    };
  }, [inView, rootMargin, threshold]);

  return (
    <div ref={ref} style={style}>
      <Suspense fallback={null}>
        <SuspendConditionally suspend={!inView}>
          {children}
        </SuspendConditionally>
      </Suspense>
    </div>
  );
};
