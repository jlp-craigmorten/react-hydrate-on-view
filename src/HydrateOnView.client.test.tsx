import { act, cleanup, render, screen } from '@testing-library/react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import type { Mock } from 'vitest';
import { HydrateOnView } from './HydrateOnView';

let mockIntersectionObserver: {
  observe: Mock;
  unobserve: Mock;
  disconnect: Mock;
};

let intersectionCallback: (entries: IntersectionObserverEntry[]) => void;

describe('HydrateOnView', () => {
  beforeEach(() => {
    vi.clearAllMocks();

    mockIntersectionObserver = {
      observe: vi.fn(),
      unobserve: vi.fn(),
      disconnect: vi.fn(),
    };

    window.IntersectionObserver = vi.fn().mockImplementation((callback) => {
      intersectionCallback = callback;

      return mockIntersectionObserver;
    });
  });

  afterEach(() => {
    cleanup();
  });

  it('should render just the container div when the component is not in the viewport', () => {
    const { container } = render(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    expect(screen.queryByTestId('child')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should observe the container div with an IntersectionObserver', () => {
    const { container } = render(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    expect(window.IntersectionObserver).toHaveBeenCalledWith(
      expect.any(Function),
      {},
    );

    expect(mockIntersectionObserver.observe).toHaveBeenCalledWith(
      container.firstChild,
    );
  });

  it('should observe the container div with a customised IntersectionObserver', () => {
    const { container } = render(
      <HydrateOnView rootMargin="10px 20px 30px 40px" threshold={0.5}>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    expect(window.IntersectionObserver).toHaveBeenCalledWith(
      expect.any(Function),
      {
        rootMargin: '10px 20px 30px 40px',
        threshold: 0.5,
      },
    );

    expect(mockIntersectionObserver.observe).toHaveBeenCalledWith(
      container.firstChild,
    );
  });

  it('should not hydrate the content if the IntersectionObserver observes an event where the content is not in the viewport', () => {
    render(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    const mockEntry = {
      isIntersecting: false,
    } as IntersectionObserverEntry;

    act(() => {
      intersectionCallback([mockEntry]);
    });

    expect(screen.queryByTestId('child')).not.toBeInTheDocument();
  });

  it('should disconnect the IntersectionObserver and hydrate the content when it enters the viewport', () => {
    const { container } = render(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    const mockEntry = {
      isIntersecting: true,
    } as IntersectionObserverEntry;

    act(() => {
      intersectionCallback([mockEntry]);
    });

    expect(mockIntersectionObserver.disconnect).toHaveBeenCalled();
    expect(screen.getByTestId('child')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should disconnect the IntersectionObserver on unmount', () => {
    const { unmount } = render(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    unmount();

    expect(mockIntersectionObserver.disconnect).toHaveBeenCalled();
  });
});
