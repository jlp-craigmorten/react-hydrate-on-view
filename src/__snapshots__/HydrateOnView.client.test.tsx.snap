// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HydrateOnView > should continue to render the provided style on the container div once hydrated 1`] = `
<div>
  <div
    style="background: red;"
  >
    <div
      data-testid="child"
    >
      Test Content
    </div>
  </div>
</div>
`;

exports[`HydrateOnView > should disconnect the IntersectionObserver and hydrate the content when it enters the viewport 1`] = `
<div>
  <div>
    <div
      data-testid="child"
    >
      Test Content
    </div>
  </div>
</div>
`;

exports[`HydrateOnView > should render just the container div when the component is not in the viewport 1`] = `<div />`;

exports[`HydrateOnView > should render the container div with custom styles 1`] = `
<div
  style="background: red;"
/>
`;
